src/server/task_manager/task.ts

```diff
@@ -0,0 +1,239 @@
+   * is treated as a single-run task, and will not be rescheduled after completion.
+   */
+  runAt?: Date;
+
+  /**
+   * If specified, indicates that the task failed to accomplish its work. This is
+   * logged out as a warning, and the task will be reattempted after a delay.
+   */
+  error?: object;
+
+  /**
+   * The state which will be passed to the next run of this task (if this is a
+   * recurring task). See the RunContext type definition for more details.
+   */
+  state?: object;
+}
+
+export const validateRunResult = Joi.object({
+  runAt: Joi.date().optional(),
+  error: Joi.object().optional(),
+  state: Joi.object().optional(),
+}).optional();
+
+export type RunFunction = () => Promise<RunResult | undefined | void>;
+
+export type CancelFunction = () => Promise<RunResult | undefined | void>;
+
+export interface CancellableTask {
+  run: RunFunction;
+  cancel?: CancelFunction;
+}
+
+export type TaskRunCreatorFunction = (context: RunContext) => CancellableTask;
+
+/**
+ * Defines a task which can be scheduled and run by the Kibana
+ * task manager.
+ */
+export interface TaskDefinition {
+  /**
+   * A unique identifier for the type of task being defined.
+   */
+  type: string;
+
+  /**
+   * A brief, human-friendly title for this task.
+   */
+  title: string;
+
+  /**
+   * An optional more detailed description of what this task does.
+   */
+  description?: string;
+
+  /**
+   * How long, in minutes, the system should wait for the task to complete
+   * before it is considered to be timed out. (e.g. '5m', the default). If
+   * the task takes longer than this, Kibana will send it a kill command and
+   * the task will be re-attempted.
+   */
+  timeOut?: string;
+
+  /**
+   * The numer of workers / slots a running instance of this task occupies.
+   * This defaults to 1.
+   */
+  numWorkers?: number;
+
+  /**
+   * Creates an object that has a run function which performs the task's work,
+   * and an optional cancel function which cancels the task.
+   */
+  createTaskRunner: TaskRunCreatorFunction;
+}
+
+/**
+ * A task definition with all of its properties set to a valid value.
+ */
+export interface SanitizedTaskDefinition extends TaskDefinition {
+  numWorkers: number;
+}
+
+export const validateTaskDefinition = Joi.object({
+  type: Joi.string().required(),
+  title: Joi.string().optional(),
+  description: Joi.string().optional(),
+  timeOut: Joi.string().default('5m'),
+  numWorkers: Joi.number().default(1),
+  createTaskRunner: Joi.func().required(),
+}).default();
+
+/**
+ * A dictionary mapping task types to their definitions.
+ */
+export interface TaskDictionary<T extends TaskDefinition> {
+  [taskType: string]: T;
+}
+
+export type TaskStatus = 'idle' | 'running';
+
+/*
+ * A task instance represents all of the data required to store, fetch,
+ * and execute a task.
+ */
+export interface TaskInstance {
+  /**
+   * The task definition type whose run function will execute this instance.
+   */
+  taskType: string;
```

Comment:
I think we said we are going to allow a hardcoded ID to be passed in, so a plugin can easily find a task it scheduled whenever it needs to. Need me to add that?